src/main/java/com/wanted/challenge/product/service/ProductService.java

```diff
@@ -0,0 +1,171 @@
+import com.wanted.challenge.product.response.PurchaseProductResponse;
+import com.wanted.challenge.product.response.ReserveProductResponse;
+import com.wanted.challenge.transact.entity.Transact;
+import com.wanted.challenge.transact.entity.TransactLog;
+import com.wanted.challenge.transact.model.TransactState;
+import com.wanted.challenge.transact.repository.TransactLogRepository;
+import com.wanted.challenge.transact.repository.TransactRepository;
+import java.util.Collections;
+import java.util.Comparator;
+import java.util.LinkedHashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.Optional;
+import java.util.stream.Collectors;
+import lombok.RequiredArgsConstructor;
+import org.springframework.data.domain.Page;
+import org.springframework.data.domain.Pageable;
+import org.springframework.stereotype.Service;
+import org.springframework.transaction.annotation.Transactional;
+
+@Service
+@RequiredArgsConstructor
+@Transactional(readOnly = true)
+public class ProductService {
+
+    private final AccountRepository accountRepository;
+    private final ProductRepository productRepository;
+    private final TransactRepository transactRepository;
+    private final TransactLogRepository transactLogRepository;
+
+    @Transactional
+    public Long register(String name, Price price, Quantity quantity, Long sellerId) {
+        Account seller = accountRepository.getReferenceById(sellerId);
+
+        Product product = new Product(seller, name, price, quantity);
+        productRepository.save(product);
+
+        return product.getId();
+    }
+
+    @Transactional
+    public void purchase(Long productId, Long buyerId) {
+        Product product = productRepository.findProductWithUpdateLockById(productId)
+                .orElseThrow(() -> new CustomException(ExceptionStatus.NOT_FOUND));
+
+        validBuyer(buyerId, product);
+        validPurchaseAlready(buyerId, product.getId());
+        validQuantity(product.getQuantity());
+
+        product.purchase();
+
+        Account buyer = accountRepository.getReferenceById(buyerId);
+        Transact transact = new Transact(buyer, product);
+        transactRepository.save(transact);
+    }
+
+    private void validBuyer(Long buyerId, Product product) {
+        if (product.getSeller().getId().equals(buyerId)) {
+            throw new CustomException(ExceptionStatus.SELF_BUY);
+        }
+    }
+
+    private void validPurchaseAlready(Long buyerId, Long productId) {
+        Optional<Transact> optTransact = transactRepository.findByBuyerIdAndProductId(buyerId, productId);
+
+        if (optTransact.isPresent()) {
+            throw new CustomException(ExceptionStatus.PURCHASE_ALREADY);
+        }
+    }
+
+    private void validQuantity(Quantity quantity) {
+        if (Objects.isNull(quantity) || quantity.value() <= 0) {
+            throw new CustomException(ExceptionStatus.CAN_NOT_PURCHASE);
+        }
+    }
```

Comment:
```suggestion
    @Transactional
    public void purchase(Long productId, Long buyerId) {
        Product product = productRepository.findProductWithUpdateLockById(productId)
                .orElseThrow(() -> new CustomException(ExceptionStatus.NOT_FOUND));

        product.purchase(buyerId);

        Account buyer = accountRepository.getReferenceById(buyerId);
        Transact transact = new Transact(buyer, product);
        transactRepository.save(transact);
    }
```
어차피 상품을 구매하는데 필요한 검증로직이라면 Product entity 내부에 검증로직을 넣는 것도 좋은 방법인 것 같습니다. Service layer가 더 간결해지고, 기타 다른 servicer에서 추가로 검증로직을 구현할 수고를 덜어줄 것 같아요. 또 검증에 필요한 값들이 어차피 product 안에 있기 때문에 검증로직을 service layer까지 올릴 필요는 없는 것 같습니다.

Reply:
`validBuyer`, `validQuantity`는 Product 내부에 있어도 손색이 없을 것 같네요..! 한번 수정해 보도록 하겠습니다!
다만 `validPurchaseAlready`의 경우 `transactRepository`를 사용하다 보니 Entity 내부에 넣기가 힘들 것 같네요.. Transact와 Product도 서로 다른 영역이기도 하구요..

해당 검증은 서비스에 두는 게 더 나을까요? 혹은 더 나은 해결 방안이 있을까요..?

Reply:
정답은 없지만 저는 이런 경우에 중복구매를 체크하는 도메인 서비스를 별도로 만들어 처리하는 편입니다.
검증을 service에서 하고 싶으시면 Product에 purchase() 메서드 이름을 (예를 들어서 discountStock로) 바꾸는 게  혼동이 없을 것 같아요.